File: lib/services/finnhub_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:stocks_app/models/stock.dart';
import 'package:stocks_app/models/news.dart';

class FinnhubService {
  static const String _baseUrl = 'https://finnhub.io/api/v1';
  static const String _apiKey = 'ctf5eu9r01qngiddatvgctf5eu9r01qngiddau00';

  Future<Stock?> getStockQuote(String symbol) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/quote?symbol=$symbol&token=$_apiKey'),
      );
      
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return Stock(
          symbol: symbol,
          currentPrice: data['c'].toDouble(),
          change: data['d'].toDouble(),
          percentChange: data['dp'].toDouble(),
        );
      }
    } catch (e) {
      print('Error fetching stock quote: $e');
    }
    return null;
  }

  Future<List<News>> getMarketNews() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/news?category=general&token=$_apiKey'),
      );
      
      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        return data.take(5).map((item) => News.fromJson(item)).toList();
      }
    } catch (e) {
      print('Error fetching market news: $e');
    }
    return [];
  }
}

File: src/services/finnhubService.ts
import axios from 'axios';
import { FINNHUB_API_KEY, FINNHUB_BASE_URL } from '../utils/constants';
import { Stock, News } from '../types';

export const getStockQuote = async (symbol: string): Promise<Stock | null> => {
  try {
    const response = await axios.get(
      `${FINNHUB_BASE_URL}/quote?symbol=${symbol}&token=${FINNHUB_API_KEY}`
    );
    
    return {
      symbol,
      currentPrice: response.data.c,
      change: response.data.d,
      percentChange: response.data.dp,
    };
  } catch (error) {
    console.error('Error fetching stock quote:', error);
    return null;
  }
};

export const getMarketNews = async (): Promise<News[]> => {
  try {
    const response = await axios.get(
      `${FINNHUB_BASE_URL}/news?category=general&token=${FINNHUB_API_KEY}`
    );
    
    return response.data.slice(0, 5).map((item: any) => ({
      headline: item.headline || '',
      summary: item.summary || '',
      url: item.url || '',
      source: item.source || '',
    }));
  } catch (error) {
    console.error('Error fetching market news:', error);
    return [];
  }
};